include/alpaka/block/shared/st/BlockSharedMemStMember.hpp

```diff
@@ -54,12 +54,12 @@ namespace alpaka
                         void alloc() const
                         {
                             m_allocdBytes = allocPitch<T>();
+#if (defined ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST) && (! defined NDEBUG)
+                            ALPAKA_ASSERT(m_allocdBytes + sizeof(T) <= m_capacity);
+#endif
                             uint8_t* buf = &m_mem[m_allocdBytes];
                             new (buf) T();
```

Comment:
I was actually thinking about removing this placement new, because `alloc()` did not initialize the memory in any backend. While type allocated here should be default constructible, in some exotic cases default constructing causes problems because of compiler bugs (e.g. generated default ctor calls memset()).

Reply:
Maybe we could just remove the `new` and not move the assert.

Reply:
Maybe, I do not have a particular opinion regarding `new`.
But I think my change has to be applied anyways, as already `m_mem[m_allocdBytes]` is potentially dangerous and should happen after the debug check.

Reply:
Ahh, you are referring to the case `m_mem == nullptr`. Maybe this should be checked in the `BlockSharedMemStMember` ctor explicitly. The assert here relies on `m_capacity` telling the truth, which we should no rely on if we want to catch bugs like this.

Reply:
Regarding `new` it is pointless here, because it is not required and might cause issues which can be avoided by just not having it.

(Btw. I cannot edit here :( )

Reply:
Or maybe we do not want to check in the ctor, if the assume passing `nullptr` to be valid usage, where you are then just not allowed to call malloc. In this case your check should be enough.

Reply:
Okay, I implemented suggestion of @jkelling and removed the placement `new`.